sites/kit.svelte.dev/src/lib/search/SearchBox.svelte

```diff
@@ -0,0 +1,416 @@
+		const { blocks } = await response.json();
+
+		index = new flexsearch.Index({
+			tokenize: 'forward'
+		});
+
+		lookup = new Map();
+
+		let time = Date.now();
+		for (const block of blocks) {
+			const title = block.breadcrumbs[block.breadcrumbs.length - 1];
+			lookup.set(block.href, {
+				title,
+				href: block.href,
+				breadcrumbs: block.breadcrumbs.slice(0, -1),
+				content: block.content
+			});
+			index.add(block.href, `${title} ${block.content}`);
+
+			// poor man's way of preventing blocking
+			if (Date.now() - time > 25) {
+				await new Promise((f) => setTimeout(f, 0));
+				time = Date.now();
+			}
+		}
+
+		update();
+	});
+
+	afterNavigate(() => {
+		// TODO this also needs to apply when only the hash changes
+		// (should before/afterNavigate fire at that time? unclear)
+		$searching = false;
+	});
+
+	function update() {
+		results = (index ? index.search($query) : []).map((href) => lookup.get(href));
+	}
+
+	function escape(text) {
+		return text.replace(/</g, '&lt;').replace(/>/g, '&gt;');
+	}
+
+	function excerpt(content, query) {
+		const index = content.toLowerCase().indexOf(query.toLowerCase());
+		if (index === -1) {
+			return content.slice(0, 100);
+		}
+
+		const prefix = index > 20 ? `…${content.slice(index - 15, index)}` : content.slice(0, index);
+		const suffix = content.slice(
+			index + query.length,
+			index + query.length + (80 - (prefix.length + query.length))
+		);
+
+		return (
+			escape(prefix) +
+			`<mark>${escape(content.slice(index, index + query.length))}</mark>` +
+			escape(suffix)
+		);
+	}
+
+	/** @param {string} href */
+	function navigate(href) {
+		$recent = [href, ...$recent.filter((x) => x !== href)];
+		$searching = false;
+	}
+
+	$: if ($searching) update();
+
+	$: recent_searches = lookup ? $recent.map((href) => lookup.get(href)).filter(Boolean) : [];
+</script>
+
+<svelte:window
+	on:keydown={(e) => {
+		if (e.code === 'KeyK' && (navigator.platform === 'MacIntel' ? e.metaKey : e.ctrlKey)) {
+			e.preventDefault();
+			$query = '';
+			$searching = !$searching;
+		}
+
+		if (e.code === 'Escape') {
+			$searching = false;
+		}
+	}}
+/>
+
+{#if $searching && index}
+	<div
+		class="modal-background"
+		on:click={() => ($searching = false)}
+		on:wheel={(e) => e.preventDefault()}
+		on:touchmove={(e) => e.preventDefault()}
+	/>
+
+	<div
+		bind:this={modal}
+		class="modal"
+		on:keydown={(e) => {
+			if (e.key === 'ArrowDown' || e.key === 'ArrowUp') {
+				e.preventDefault();
+				const group = focusable_children(e.currentTarget);
+
+				// when using arrow keys (as opposed to tab), don't focus buttons
+				const selector = 'a, input';
+
+				if (e.key === 'ArrowDown') {
+					group.next(selector);
+				} else {
+					group.prev(selector);
+				}
+			}
+
+			if (e.code === 'Space' && e.target.nodeName !== 'INPUT') {
+				e.preventDefault();
+			}
+		}}
+		use:trap
+	>
+		<div class="search-box">
+			<!-- svelte-ignore a11y-autofocus -->
+			<input
```

Comment:
After further reflection, I don't think this is a combobox or autocomplete (as I suggested in another comment), since we aren't having the user select a value. It's just an input that dynamically filters a list of search results. With that in mind, I think we can apply some recommendations from ["Considering dynamic search results and content"](https://www.scottohara.me/blog/2022/02/05/dynamic-results.html) by Scott O'Hara (posted today!)

- Add [instructions](https://www.scottohara.me/blog/2022/02/05/dynamic-results.html#the-text-field) that say something like "Results will appear as you type", linked to the input with aria-describedby. These could be visually hidden so they're only shown to assistive tech users, if desired.
- Add assertive live region to [announce when no results are found](https://www.scottohara.me/blog/2022/02/05/dynamic-results.html#communicating-no-results)

Reply:
thanks — have implemented both of this